### ISSUE TITLE
Answering call with callkit when App is in background and the phone is locked, there is no audio.

### ISSUE THREAD

**Issue by sanketns** [livekit/client-sdk-swift#181]
when the app is in background and the phone is locked. There is no audio of opponent user. Tried every possible way but no luck. Kindly guide me for the same.

**Comment by hiroshihorie** [comment]
Hi @sanketns I would be happy to investigate, if you could prepare a **minimal reproducible example** simulating your issue.

**Comment by sanketns** [comment]
Getting these errors in console randomly, see if you can get any hints from this if possible.

ATAudioSessionPropertyManager.mm:359  FAILED to set property 1684431725 on AVAudioSession with error Error Domain=NSOSStatusErrorDomain Code=-50 "(null)"

AURemoteIO.cpp:1679 AUIOClient_StartIO failed (561015905)

**Comment by raphaelalkamim** [comment]
Hello, I’m experiencing the same issue. When I create a room (in my case, start a call), if I immediately go to the background without waiting for the video/audio section to be initialized, I encounter these errors:

**AURemoteIO.cpp:2087 AURemoteIO@0x10b025c40: IOThread exiting with error 0x10004006
AURemoteIO.cpp:1702 AUIOClient_StartIO failed (561145187)** 

Otherwise, if I wait a couple of seconds and the audio/video initializes before going to the background, everything works fine.

**Comment by sanketns** [comment]
@raphaelalkamim , may be it's related to audio session configuration while making a call. By the way, are you doing this with callkit?

**Comment by raphaelalkamim** [comment]
@sanketns , Yes, you're right, it might be related to the audio session configuration during calls. I'm currently using the default LiveKit configuration and CallKit solely for managing the call display.

**Comment by sanketns** [comment]
@raphaelalkamim , Are you doing something when these below functions get called?

func provider(_ provider: CXProvider, didActivate audioSession: AVAudioSession){
  
}
func provider(_ provider: CXProvider, didDeactivate audioSession: AVAudioSession) {
 
}

**Comment by raphaelalkamim** [comment]
@sanketns  No, I'm not using these delegates at all. Should I configure something beyond the LiveKit default configuration?

**Comment by ajayz7** [comment]
Hi any solution to fix this . I am experiencing same when using callKit? @hiroshihorie @sanketns

**Comment by dobreandl** [comment]
Hello, we've encountered the same issue.

To reproduce this, you need to create a CallKit app and use the LiveKit SDK for audio streaming.

The issue happens when you receive a call and the device is locked. 

When you're opening the app and creating the WebRTC session, there is audio, but the microphone is not working. This is a common issue in the WebRTC it seems, and a new WebRTC session needs to be created, however, I'm not sure how to do this with the LiveKit SDK.

This SO thread describes the issue https://stackoverflow.com/questions/45366289/ios-audio-not-working-during-call-answered-when-phone-is-locked-webrtc-used-for

@sanketns @hiroshihorie Could you advise?

**Comment by liuyang2211** [comment]
Yes, I have the same situation. Is there any solution

**Comment by zackzhuang1996** [comment]
alright according to the team, the iOS SDK doesn't support background mode. 
<img width="615" alt="Screenshot 2024-07-30 at 1 22 33 AM" src="https://github.com/user-attachments/assets/2e9cff31-f919-4888-a003-7cd08cdbec42">

**Comment by davidliu** [comment]
@zackzhuang1996 that's the react native sdk, which is separate from this swift sdk.

@sanketns
For CallKit, the `didActivate` and `didDeactivate` methods should be implemented as follows:

1. Add `import LiveKitWebRTC` to your CallProvider file.

2. 
```
func provider(_ provider: CXProvider, didActivate audioSession: AVAudioSession){
    LKRTCAudioSession.sharedInstance().audioSessionDidActivate(audioSession)
}
func provider(_ provider: CXProvider, didDeactivate audioSession: AVAudioSession) {    
    LKRTCAudioSession.sharedInstance().audioSessionDidDeactivate(audioSession)
}
```

**Comment by Viktorianec** [comment]
I have the same error when I answer with callkit. Also I received this message:
`AudioManager.defaultConfigureAudioSessionFunc(newState:oldState:) Failed to configure audio session with error: Error Domain=NSOSStatusErrorDomain Code=1701737535 "Session activation failed" UserInfo={NSLocalizedDescription=Session activation failed}`

**Comment by hanyibo520** [comment]
I have the same error when I answer with callkit. How did everyone handle it?

**Comment by saadsaadi** [comment]
Any updates on this?

**Comment by saadsaadi** [comment]
> I have the same error when I answer with callkit. Also I received this message: `AudioManager.defaultConfigureAudioSessionFunc(newState:oldState:) Failed to configure audio session with error: Error Domain=NSOSStatusErrorDomain Code=1701737535 "Session activation failed" UserInfo={NSLocalizedDescription=Session activation failed}`
> 
> EDIT:
> 
> On the first call, there is incoming audio in the background. There is no incoming audio on the second and subsequent calls. At the same time, the other party can hear me all the time.
> 
> Could be problem here: `audioSessionDidDeactivate`?

@Viktorianec did you happen to solve this?

**Comment by saadsaadi** [comment]
I have this workaround.

```
extension RemoteRoom: RoomDelegate {

  nonisolated func room(
     _ room: Room,
     participant: RemoteParticipant,
     didPublishTrack publication: RemoteTrackPublication
   ) {
    if audioTrackRefreshed { return }
    audioTrackRefreshed = true

    do {
      for track in room.localParticipant.localAudioTracks {
        try await room.localParticipant.unpublish(publication: track)
      }

      let track = LocalAudioTrack.createTrack(name: "microphone")
      try await room.localParticipant.publish(audioTrack: track)
      try await setIsMicEnabled(room: room, val: isMicEnabled)
    } catch {
      print("Error publishing audio track: \(error)")
    }
}
```

I wish I didn't have to do this but I can't figure out how to get it properly working.

**Comment by seyipaye** [comment]
Thank you for sharing @saadsaadi. I've been thinking of how to resolve this, and you shared the solution just now 👍

**Comment by seyipaye** [comment]
Please can you confirm with me if "audioTrackRefreshed" was initially declared as True or False?

**Comment by seyipaye** [comment]
Hi @saadsaadi. Does your solution work for you?

I just tested it briefly, and sadly, the problem still persists

**Comment by mandic1987** [comment]
I have the same issue on version 2.0.16... Is there any solution to this problem?

**Comment by hiroshihorie** [comment]
I think it's a timing issue of starting the internal AVAudioEngine:

```swift
// Keep this true (Default)
AudioManager.shared.audioSession.isAutomaticConfigurationEnabled = true
```

```swift
// Disable auto subscribe, since this will try to configure/start audio engine right after room connects.
Connect to room with autoSubscribe = false
```

I think the key is to configure AVAudioSession & start internal AVAudioEngine at `provider(_:, didActivate:)` since that's the only timing AVAudioSession active & the SDK needs it to be configured otherwise will fail to start audio engine.

```swift
func provider(_ provider: CXProvider, didActivate audioSession: AVAudioSession) {
  // Subscribe to your remote track here, or publish mic which will configure audio session and start audio engine.
}

func provider(_ provider: CXProvider, didDeactivate audioSession: AVAudioSession) {    
  // Un-subscribe remote tracks / un-publish mic which will deactivate audio session and turn off internal audio engine.
}
```

For new SDK versions I don't touch the `LKRTCAudioSession` at all (in the AVAudioEngine version AudioDeviceModule) so I don't think you need to:
```swift
LKRTCAudioSession.sharedInstance().audioSessionDidActivate(audioSession) // Not required
```

I hope we can provide a sample app for this.

**Comment by pblazej** [comment]
@hiroshihorie re-assigned that to you, as you already started digging into that 🥇

**Comment by m-ricca** [comment]
I encountered the same error and resolved it by doing the following:
first: `room.connect` and, once connected to the room, in the `didActivate` of the `audioSession` do:
`LKRTCAudioSession.sharedInstance().audioSessionDidActivate(audioSession)`
and then activate the room's audio and microphone with:
```
try await room.localParticipant.setCamera(enabled: false)
try await room.localParticipant.setMicrophone(enabled: true)
try await room.localParticipant.setScreenShare(enabled: false)
```

**Comment by hiroshihorie** [comment]
Update: I'm working on a small CallKit example and a PR (https://github.com/livekit/client-sdk-swift/pull/786) to help coordinate timing of AVAudioSession configuration & AVAudioEngine start / stop within the `didActivate` ~ `didDeactivate` window.

**Comment by hiroshihorie** [comment]
Update: The [example app](https://github.com/livekit-examples/swift-example-collection/tree/hiroshi/callkit/callkit#readme) is now complete.

https://github.com/user-attachments/assets/c1ce1dbd-74c0-40aa-9018-30c2829aa55c

This relies on https://github.com/livekit/client-sdk-swift/pull/786 and we will release it after more testing.

Please read the [README of the example](https://github.com/livekit-examples/swift-example-collection/tree/hiroshi/callkit/callkit#readme) for details.

These are issues I've encountered while implementing this example. Please let me know if there are better workarounds or if I'm incorrect.

1. [reportNewIncomingCall(with:update:completion:)](https://developer.apple.com/documentation/callkit/cxprovider/reportnewincomingcall(with:update:completion:)) must be invoked on the same thread as [pushRegistry(_:didReceiveIncomingPushWith:for:completion:)](https://developer.apple.com/documentation/pushkit/pkpushregistrydelegate/pushregistry(_:didreceiveincomingpushwith:for:completion:)), otherwise it will silently fail and not show the incoming call UI on the lock screen. Therefore, the async version may not work as intended.

2. You may need to set AVAudioSession.category to .playAndRecord before calling [reportNewIncomingCall(with:update:completion:)](https://developer.apple.com/documentation/callkit/cxprovider/reportnewincomingcall(with:update:completion:)) when your app is woken up in the background by [pushRegistry(_:didReceiveIncomingPushWith:for:completion:)](https://developer.apple.com/documentation/pushkit/pkpushregistrydelegate/pushregistry(_:didreceiveincomingpushwith:for:completion:)).

I find Issue 2 ☝️ really odd since the session configuration should be handled at [provider(_:didActivate:)](https://developer.apple.com/documentation/callkit/cxproviderdelegate/provider(_:didactivate:)) and [provider(_:didDeactivate:)](https://developer.apple.com/documentation/callkit/cxproviderdelegate/provider(_:diddeactivate:)).

These were hard / time-consuming issues to figure out, IMO Apple should document it well or fix these problems since a lot of people seem to run into issues.

Related: https://developer.apple.com/forums/thread/783870

**Comment by alon7** [comment]
@hiroshihorie, wonderful news. Going to test this today!

**Comment by alon7** [comment]
@hiroshihorie I see #786 was merged, any plans on releasing 2.7.3 soon for easier testing?

**Comment by hiroshihorie** [comment]
@alon7 I will release a new version after #774 #793 are merged in 🙂 meanwhile you can test if you depend on `main` branch using SPM.

**Comment by alon7** [comment]
Thanks @hiroshihorie. Just did that and it worked. We have a bit of a unique situation where I need to use cocoapods and this repo doesn't support newer versions, but it ended up working. Testing and Will keep you posted!

**Comment by pnicholls** [comment]
@hiroshihorie is it ok to turn the microphone on therefore request mic permission in the background when I call comes in but has yet to be answered?

**Comment by akaraul** [comment]
> [@alon7](https://github.com/alon7) I will release a new version after [#774](https://github.com/livekit/client-sdk-swift/pull/774) [#793](https://github.com/livekit/client-sdk-swift/pull/793) are merged in 🙂 meanwhile you can test if you depend on `main` branch using SPM.

It seems this is already merged. Please release a new version.

**Comment by hiroshihorie** [comment]
@pnicholls If you first:
```swift
try AudioManager.shared.setEngineAvailability(.none)
```
Even you publish/enable the mic track, mic perms won't be requested until you:
```swift
try AudioManager.shared.setEngineAvailability(.default)
```

Pleas see example for details.

**Comment by hiroshihorie** [comment]
Hello, v2.8.1 is out now and the example app will work with this version.
